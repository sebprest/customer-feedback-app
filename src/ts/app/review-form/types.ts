import { Review } from "^/common/types";
import { ADD_REVIEW } from "./actions";

export type AddReviewAction = {
  type: typeof ADD_REVIEW;
  payload: Review;
};

export type FormState = {
  name: string;
  email: string;
  rating: number;
  comment: string;
};
