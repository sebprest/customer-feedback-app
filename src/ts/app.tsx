import React from "react";

const App: React.FunctionComponent = React.memo(() => <h1>Hello</h1>);

export default App;
