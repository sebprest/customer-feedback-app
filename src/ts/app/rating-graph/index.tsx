import { Line, LineChart, ResponsiveContainer, YAxis } from "recharts";
import React from "react";
import { connect } from "react-redux";

import { ReviewsState } from "^/common/types";
import { StoreState } from "^/store/types";

interface StateProps {
  reviews: ReviewsState;
}

type Props = StateProps;

export const RatingGraph: React.FunctionComponent<Props> = React.memo(
  (props) => {
    return (
      <div className="rating-graph">
        <ResponsiveContainer>
          <LineChart data={props.reviews}>
            <YAxis />
            <Line type="monotone" dataKey={"rating"} animationDuration={0} />
          </LineChart>
        </ResponsiveContainer>
      </div>
    );
  }
);

const mapStateToProps = ({ reviews }: StoreState): StateProps => {
  return {
    reviews,
  };
};

export default connect(mapStateToProps)(RatingGraph);
