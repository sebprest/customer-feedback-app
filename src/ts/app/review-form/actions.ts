import { Review } from "^/common/types";
import { AddReviewAction } from "./types";

export const ADD_REVIEW = "ADD_REVIEW";
export function addReview(newReview: Review): AddReviewAction {
  return {
    type: ADD_REVIEW,
    payload: newReview,
  };
}
