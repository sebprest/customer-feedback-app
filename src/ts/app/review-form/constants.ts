export const MINIMUM_REVIEW_RATING = 1;
export const MAXIMUM_REVIEW_RATING = 5;

export const REVIEW_FORM_FIELD_IDS = {
  name: "review-form-name",
  email: "review-form-email",
  rating: "review-form-rating",
  comment: "review-form-comment",
};
