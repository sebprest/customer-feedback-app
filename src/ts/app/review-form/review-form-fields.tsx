import React, { useState } from "react";

import { Review } from "^/common/types";
import {
  MAXIMUM_REVIEW_RATING,
  MINIMUM_REVIEW_RATING,
  REVIEW_FORM_FIELD_IDS,
} from "./constants";

type Props = {
  addReview: (newReview: Review) => void;
};

export const ReviewFormFields: React.FunctionComponent<Props> = React.memo(
  (props) => {
    const defaultValues = { name: "", email: "", rating: 0, comment: "" };
    const [values, setValues] = useState(defaultValues);

    const handleInputChange = (event: React.ChangeEvent<HTMLInputElement>) => {
      const targetName = event.target.name;
      const targetValue = event.target.value;

      setValues((values) => ({ ...values, [targetName]: targetValue }));
    };

    const onAddReview = (event: React.FormEvent<HTMLFormElement>) => {
      event.preventDefault();
      props.addReview(values);
      setValues(() => defaultValues);
    };

    return (
      <form className="review-form" onSubmit={onAddReview}>
        <label htmlFor={REVIEW_FORM_FIELD_IDS.name}>Name</label>
        <input
          id={REVIEW_FORM_FIELD_IDS.name}
          type="text"
          name="name"
          value={values.name}
          onChange={handleInputChange}
          required
        />
        <label htmlFor={REVIEW_FORM_FIELD_IDS.email}>Email</label>
        <input
          id={REVIEW_FORM_FIELD_IDS.email}
          type="text"
          name="email"
          value={values.email}
          onChange={handleInputChange}
          required
        />
        <label htmlFor={REVIEW_FORM_FIELD_IDS.rating}>Rating</label>
        <input
          type="number"
          id={REVIEW_FORM_FIELD_IDS.rating}
          name="rating"
          value={values.rating}
          onChange={handleInputChange}
          required
          min={MINIMUM_REVIEW_RATING}
          max={MAXIMUM_REVIEW_RATING}
        />
        <label htmlFor={REVIEW_FORM_FIELD_IDS.comment}>Comment</label>
        <input
          id={REVIEW_FORM_FIELD_IDS.comment}
          name="comment"
          value={values.comment}
          onChange={handleInputChange}
          required
        />
        <input className="button" type="submit" value="Submit" />
      </form>
    );
  }
);

export default ReviewFormFields;
