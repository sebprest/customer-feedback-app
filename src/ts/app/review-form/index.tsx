import React, { useState } from "react";
import { connect } from "react-redux";

import { addReview } from "./actions";
import ReviewFormFields from "./review-form-fields";

type Props = {
  addReview: typeof addReview;
};

export const ReviewForm: React.FunctionComponent<Props> = React.memo(
  (props) => {
    const defaultValues = { name: "", email: "", rating: 0, comment: "" };
    const [values, setValues] = useState(defaultValues);

    const handleInputChange = (event: React.ChangeEvent<HTMLInputElement>) => {
      const targetName = event.target.name;
      const targetValue = event.target.value;

      setValues((values) => ({ ...values, [targetName]: targetValue }));
    };

    const onAddReview = (event: React.FormEvent<HTMLFormElement>) => {
      event.preventDefault();
      props.addReview(values);
      setValues(() => defaultValues);
    };

    return (
      <ReviewFormFields
        addReview={onAddReview}
        updateFormState={handleInputChange}
        formState={values}
      />
    );
  }
);

export default connect(null, { addReview })(ReviewForm);
