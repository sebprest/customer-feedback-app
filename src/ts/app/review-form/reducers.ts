import { Review, ReviewsState } from "^/common/types";
import { ADD_REVIEW } from "./actions";
import { AddReviewAction } from "./types";

export function review(
  state: Review[] = [],
  action: AddReviewAction
): ReviewsState {
  switch (action.type) {
    case ADD_REVIEW:
      return [...state, action.payload];
    default:
      return state;
  }
}
