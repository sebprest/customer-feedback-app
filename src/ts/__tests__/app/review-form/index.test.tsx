jest.mock("^/app/review-form/review-form-fields", () => (props: unknown) => (
  <div {...props}>ReviewFormFields</div>
));

import React from "react";
import renderer from "react-test-renderer";

import { ReviewForm } from "^/app/review-form";

describe("ReviewForm", () => {
  it("should render the review form fields with the addReview action passed to it", () => {
    const reviewForm = renderer.create(<ReviewForm addReview={jest.fn()} />);

    expect(reviewForm).toMatchSnapshot();
  });
});
