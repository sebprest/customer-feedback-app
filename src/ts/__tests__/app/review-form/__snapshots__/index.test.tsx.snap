// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReviewForm should render the review form fields with the addReview action passed to it 1`] = `
<div
  addReview={[MockFunction]}
>
  ReviewFormFields
</div>
`;
